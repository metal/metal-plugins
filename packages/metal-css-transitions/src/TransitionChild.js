'use strict';

import Anim from 'metal-anim';
import dom from 'metal-dom';
import Component, {Config} from 'metal-jsx';

export const DELAY_TIME = 10;

/**
 * TransitionChild component
 */
class TransitionChild extends Component {
	/**
	 * @inheritDoc
	 */
	created() {
		this.transition_ = this.transition_.bind(this);
	}

	/**
	 * Handles `appear` type css transition.
	 */
	appear() {
		this.transition_('appear');
	}

	/**
	 * Handles adding the active class name after a short delay.
	 * @param {string} className Classname to apply to node.
	 * @param {!Object} node DOM node that class is applied to.
	 * @protected
	 */
	delayActive_(className, node) {
		setTimeout(() => {
			node.classList.add(className);

			Anim.emulateTransitionEnd(node);
		}, DELAY_TIME);
	}

	/**
	 * Handles `enter` type css transition.
	 */
	enter() {
		this.transition_('enter');
	}

	/**
	 * Handles `leave` type css transition.
	 * @param {function()} callback Function to run after transition.
	 */
	leave(callback) {
		this.transition_('leave', callback);
	}

	/**
	 * Handles adding the transition classname to the component element.
	 * @param {string} type Classname to apply to node.
	 * @param {function()} finishCallback Callback method to execute after transition.
	 * @protected
	 */
	transition_(type, finishCallback) {
		const node = this.element;
		if (!node) {
			return;
		}

		const className = `${this.props.name}-${type}`;
		const activeClassName = `${className}-active`;

		const callback = () => {
			if (finishCallback) {
				finishCallback();
			} else {
				node.classList.remove(activeClassName, className);
			}
		};

		dom.once(node, 'transitionend', callback);

		node.classList.add(className);

		this.delayActive_(activeClassName, node);
	}

	/**
	 * @inheritDoc
	 */
	render() {
		return this.props.children[0];
	}
}

TransitionChild.PROPS = {
	/**
	 * Name of transition.
	 * @type {string}
	 */
	name: Config.string()
};

export default TransitionChild;
