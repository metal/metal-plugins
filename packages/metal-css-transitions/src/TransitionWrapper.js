'use strict';

import Component, {Config} from 'metal-jsx';

import TransitionChild from './TransitionChild';
import {getChildrenMap, mergeChildrenMap} from './utils';

/**
 * TransitionWrapper component
 */
class TransitionWrapper extends Component {
	/**
	 * @inheritDoc
	 */
	created() {
		this.state.childrenMap_ = getChildrenMap(this.props.children);

		this.handleChildrenEnter_ = this.handleChildrenEnter_.bind(this);
		this.handleChildrenLeave_ = this.handleChildrenLeave_.bind(this);
	}

	/**
	 * Executes the `appear` method on each child.
	 */
	attached() {
		if (this.props.children.length) {
			this._justAppeared = true;
		}

		this.props.children.forEach(child => {
			if (child && child.props) {
				const {key} = child.props;

				if (key && this.refs[key]) {
					this.refs[key].appear();
				}
			}
		});
	}

	/**
	 * Removes child from `childrenMap_`.
	 * @param {string|number} key The unique identifier of the child element that left.
	 * @protected
	 */
	finishLeave_(key) {
		const newChildrenMap = this.state.childrenMap_;

		newChildrenMap.delete(key);

		this.state.childrenMap_ = newChildrenMap;
	}

	/**
	 * Executes the `enter` method on each new entering child. However, these calls
	 * will be skipped if this is the first transition since `appear` was called.
	 * @param {!Array} newChildren Children elements that are entering.
	 * @param {!Object} prevKeyMap Map of children who are already present.
	 * @protected
	 */
	handleChildrenEnter_(newChildren, prevKeyMap) {
		if (this._justAppeared) {
			this._justAppeared = false;
			return;
		}

		newChildren.forEach(child => {
			if (child && child.props) {
				const {key} = child.props;

				if (key && !prevKeyMap.has(key) && this.refs[key]) {
					this.refs[key].enter();
				}
			}
		});
	}

	/**
	 * Executes the `leave` method on each child that leaves.
	 * @param {!Array} prevChildren Children who are already present.
	 * @param {!Object} newKeyMap Map of new children.
	 * @protected
	 */
	handleChildrenLeave_(prevChildren, newKeyMap) {
		prevChildren.forEach(child => {
			if (child && child.props) {
				const {key} = child.props;

				if (key && !newKeyMap.has(key) && this.refs[key]) {
					this.refs[key].leave(this.finishLeave_.bind(this, key));
				}
			}
		});
	}

	/**
	 * Sorts children, creates new `childrenMap_` and passes them to
	 * `handleChildrenEnter_` and `handleChildrenLeave_`.
	 * @param {!Array} newChildren New children elements.
	 * @param {!Array} prevChildren Previous children elements.
	 */
	syncChildren(newChildren, prevChildren = []) {
		const newKeyMap = getChildrenMap(newChildren);
		const prevKeyMap = getChildrenMap(prevChildren);

		this.setState(
			{
				childrenMap_: mergeChildrenMap(newKeyMap, prevKeyMap)
			},
			() => {
				this.handleChildrenEnter_(newChildren, prevKeyMap);
				this.handleChildrenLeave_(prevChildren, newKeyMap);
			}
		);
	}

	/**
	 * @inheritDoc
	 */
	render() {
		const children = [];

		const {childrenMap_} = this.state;

		childrenMap_.forEach(child => {
			if (child) {
				children.push(child);
			}
		});

		return (
			<span>
				{!!children.length &&
					children.map(child => (
						// eslint-disable-next-line
						<TransitionChild
							name={this.props.name}
							ref={child.props.key}
						>
							{child}
						</TransitionChild>
					))}
			</span>
		);
	}
}

TransitionWrapper.PROPS = {
	/**
	 * Name of css transition.
	 * @type {string}
	 */
	name: Config.string()
};

TransitionWrapper.STATE = {
	/**
	 * Map of each child to its corresponding key value.
	 * @type {object}
	 * @default {}
	 * @protected
	 */
	childrenMap_: Config.value({})
};

export default TransitionWrapper;
