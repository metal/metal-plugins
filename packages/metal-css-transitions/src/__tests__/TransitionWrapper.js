'use strict';
import Component, {Config} from 'metal-jsx';

import TransitionWrapper from '../TransitionWrapper';
import TransitionChild from '../TransitionChild';

describe('TransitionWrapper', function() {
	let component;

	afterEach(() => {
		if (component) {
			component.dispose();
		}
	});

	it('renders', () => {
		component = new TransitionWrapper();

		expect(component);
	});

	it('renders without children', () => {
		class App extends Component {
			render() {
				return <TransitionWrapper elementClasses="wrapper" />;
			}
		}

		component = new App();

		expect(component).toMatchSnapshot();
	});

	it('renders with one child', () => {
		class App extends Component {
			render() {
				return (
					<TransitionWrapper name="test">
						<div class="child">Child</div>
					</TransitionWrapper>
				);
			}
		}

		component = new App();

		expect(component).toMatchSnapshot();
	});

	it('renders with multiple children', () => {
		class App extends Component {
			render() {
				return (
					<TransitionWrapper name="test">
						<div class="child1" key={1}>Child</div>
						<div class="child2" key={2}>Child</div>
					</TransitionWrapper>
				);
			}
		}

		component = new App();

		expect(component).toMatchSnapshot();
	});

	it('should call setState when new children are supplied', () => {
		const spy = jest.fn();

		component = new TransitionWrapper();

		component.setState = spy;

		expect(spy).not.toBeCalled();
		component.props.children = [{}];

		jest.runAllTimers();

		expect(spy).toBeCalled();
	});

	it('should call syncChildren with empty args', () => {
		const spy = jest.fn();

		component = new TransitionWrapper();

		component.setState = spy;

		expect(spy).not.toBeCalled();

		component.props.children = [];

		jest.runAllTimers();

		expect(spy).toBeCalled();
	});

	it('should process entering children and call .enter()', () => {
		component = new TransitionWrapper();

		const spy = jest.fn();
		const KEY = 1;

		component.refs[KEY] = {
			enter: spy
		};

		const children = [
			{
				props: {
					key: KEY
				}
			}
		];

		expect(spy).not.toBeCalled();

		component.handleChildrenEnter_(children, new Map());

		expect(spy).toBeCalled();
	});

	it('should process exiting children and call .leave()', () => {
		component = new TransitionWrapper();

		const spy = jest.fn(callback => callback());
		const KEY = 1;

		component.refs[KEY] = {
			leave: spy
		};

		const children = [
			{
				props: {
					key: KEY
				}
			}
		];

		expect(spy).not.toBeCalled();

		component.handleChildrenLeave_(children, new Map());

		expect(spy).toBeCalled();
	});

	it('should remove key from childrenMap_ state', () => {
		component = new TransitionWrapper();
		const KEY = 1;

		component.state.childrenMap_ = new Map().set(KEY, 'test');

		component.finishLeave_(KEY);

		expect(component.state.childrenMap_.has(KEY)).toBeFalsy();
	});

	it('should render children in the order they are given when keys are numbers', () => {
		class App extends Component {
			render() {
				return (
					<TransitionWrapper name="test">
						{this.props.children}
					</TransitionWrapper>
				);
			}
		}

		component = new App();

		component.props.children = [
			<div class="child" key="1">1</div>,
			<div class="child" key="2">2</div>
		];

		jest.runAllTimers();

		expect(component).toMatchSnapshot();

		component.props.children = [
			<div class="child" key={3}>3</div>,
			<div class="child" key={1}>1</div>,
			<div class="child" key={2}>2</div>
		];

		jest.runAllTimers();

		expect(component).toMatchSnapshot();
	});

	it('should render children in the order they are given when keys are strings', () => {
		class App extends Component {
			render() {
				return (
					<TransitionWrapper name="test">
						{this.props.children}
					</TransitionWrapper>
				);
			}
		}

		component = new App();

		component.props.children = [
			<div class="child" key="1">1</div>,
			<div class="child" key="2">2</div>
		];

		jest.runAllTimers();

		expect(component).toMatchSnapshot();

		component.props.children = [
			<div class="child" key="3">3</div>,
			<div class="child" key="1">1</div>,
			<div class="child" key="2">2</div>
		];

		jest.runAllTimers();

		expect(component).toMatchSnapshot();
	});

	it('should not call enter immediately after appear', () => {
		const originalEnter = TransitionChild.prototype.enter;
		const enterSpy = jest.fn(function() {
			return originalEnter.apply(this, arguments);
		});

		const originalAppear = TransitionChild.prototype.appear;
		const appearSpy = jest.fn(function() {
			return originalAppear.apply(this, arguments);
		});

		TransitionChild.prototype.enter = enterSpy;
		TransitionChild.prototype.appear = appearSpy;

		class TestComponent extends Component {
			render() {
				return (
					<TransitionWrapper>
						<div class="test-child" />
					</TransitionWrapper>
				);
			}
		}

		component = new TestComponent();

		jest.runAllTimers();

		expect(appearSpy).toBeCalled();
		expect(enterSpy).not.toBeCalled();
	});

	it('should call leave when show changes from true to false', () => {
		const originalLeave = TransitionChild.prototype.leave;

		TransitionChild.prototype.leave = jest.fn(callback => callback());

		class App extends Component {
			render() {
				return (
					<TransitionWrapper ref="transitionWrapper" name="test">
						{this.state.show_ &&
							<div class="test-child" ref="testChild">2</div>}
					</TransitionWrapper>
				);
			}
		}

		App.STATE = {
			show_: Config.bool().value(true)
		};

		component = new App();

		jest.runAllTimers();

		component.state.show_ = false;

		jest.runAllTimers();

		expect(TransitionChild.prototype.leave).toHaveBeenCalled();

		TransitionChild.prototype.leave = originalLeave;
	});

	it('should call enter when show changes from false to true', () => {
		const originalEnter = TransitionChild.prototype.enter;

		TransitionChild.prototype.enter = jest.fn();

		class App extends Component {
			render() {
				return (
					<TransitionWrapper ref="transitionWrapper" name="test">
						{this.state.show_ && <div class="test-child">2</div>}
					</TransitionWrapper>
				);
			}
		}

		App.STATE = {
			show_: Config.bool().value(false)
		};

		component = new App();

		jest.runAllTimers();

		component.state.show_ = true;

		jest.runAllTimers();

		expect(TransitionChild.prototype.enter).toHaveBeenCalled();

		TransitionChild.prototype.enter = originalEnter;
	});
});
