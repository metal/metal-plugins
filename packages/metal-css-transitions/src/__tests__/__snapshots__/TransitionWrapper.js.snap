// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransitionWrapper renders with multiple children 1`] = `
<span>
  <div class="child1 test-appear">Child</div>
  <div class="child2 test-appear">Child</div>
</span>
`;

exports[`TransitionWrapper renders with one child 1`] = `
<span>
  <div class="child test-appear">Child</div>
</span>
`;

exports[`TransitionWrapper renders without children 1`] = `<span class="wrapper"></span>`;

exports[`TransitionWrapper should render children in the order they are given when keys are numbers 1`] = `
<span>
  <div class="child">1</div>
  <div class="child">2</div>
</span>
`;

exports[`TransitionWrapper should render children in the order they are given when keys are numbers 2`] = `
<span>
  <div class="child">3</div>
  <div class="child test-leave test-leave-active">1</div>
  <div class="child test-leave test-leave-active">2</div>
</span>
`;

exports[`TransitionWrapper should render children in the order they are given when keys are strings 1`] = `
<span>
  <div class="child">1</div>
  <div class="child">2</div>
</span>
`;

exports[`TransitionWrapper should render children in the order they are given when keys are strings 2`] = `
<span>
  <div class="child">3</div>
  <div class="child">1</div>
  <div class="child">2</div>
</span>
`;
